src/query_builder.rs

```diff
@@ -0,0 +1,384 @@
+// Copyright 2018 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#![macro_use]
+use std::collections::{
+    BTreeMap,
+};
+
+use mentat_core::{
+    Entid,
+    NamespacedKeyword,
+    TypedValue,
+    ValueType,
+};
+
+use ::{
+    HasSchema,
+    Queryable,
+    QueryInputs,
+    QueryOutput,
+    Store,
+    Variable,
+};
+
+use errors::{
+    ErrorKind,
+    Result,
+};
+
+pub struct QueryBuilder<'a> {
+    sql: String,
+    values: BTreeMap<Variable, TypedValue>,
+    types: BTreeMap<Variable, ValueType>,
+    store: &'a mut Store,
+}
+
+impl<'a> QueryBuilder<'a> {
+    pub fn new<T>(store: &'a mut Store, sql: T) -> QueryBuilder where T: Into<String> {
+        QueryBuilder { sql: sql.into(), values: BTreeMap::new(), types: BTreeMap::new(), store }
+    }
+
+    pub fn bind_value<T>(mut self, var: &str, value: T) -> Self where T: Into<TypedValue> {
+        self.values.insert(Variable::from_valid_name(var), value.into());
+        self
+    }
+
+    pub fn bind_ref_from_kw(mut self, var: &str, value: NamespacedKeyword) -> Result<Self> {
+        let entid = self.store.conn().current_schema().get_entid(&value).ok_or(ErrorKind::UnknownAttribute(value.to_string()))?;
+        self.values.insert(Variable::from_valid_name(var), TypedValue::Ref(entid.into()));
+        Ok(self)
+    }
+
+    pub fn bind_ref<T>(mut self, var: &str, value: T) -> Self where T: Into<Entid> {
+       self.values.insert(Variable::from_valid_name(var), TypedValue::Ref(value.into()));
+       self
+    }
+
+    pub fn bind_instant(mut self, var: &str, value: i64) -> Self {
+       self.values.insert(Variable::from_valid_name(var), TypedValue::instant(value));
+       self
+    }
+
+    pub fn bind_type(mut self, var: &str, value_type: ValueType) -> Self {
+        self.types.insert(Variable::from_valid_name(var), value_type);
+        self
+    }
+
+    pub fn execute(&mut self) -> Result<QueryOutput> {
+        let values = ::std::mem::replace(&mut self.values, Default::default());
+        let types = ::std::mem::replace(&mut self.types, Default::default());
+        let query_inputs = QueryInputs::new(types, values)?;
+        let read = self.store.begin_read()?;
```

Comment:
If you make this function `execute(mut self)` and do

```
let (sql, values, types, store) = (self.sql, self.values, self.types, self.store);
let inputs = QueryInputs::new(types, values)?;
let read = store.begin_read()?;
read.q_once(&sql, inputs)
```

can we avoid the `mem::replace`s?

Reply:
Not if we want to maintain the non-consuming builder pattern and therefore avoid all of the boxing and unboxing in the FFI. Either, we're consuming and have an box-heavy FFI layer, in which case this is possible. Or we are non-consuming, have a smaller FFI layer that deals with references and use `mem::replace`. As far as I can tell, anyway.